### PR TITLE
WIP: memoize template compiler parse/transform/loads

### PR DESCRIPTION
**Pull request by stefanpenner**
POC fix for #331

Gotta run to a meeting, will make this not terrible shortly.

### CODE REVIEW

File: packages/core/src/template-compiler.ts [hunk 1]
```diff
@@ -41,40 +41,50 @@ interface GlimmerSyntax {
   cacheKey: string;
 }
 
+const CACHE = new Map();
```
Comment (stefanpenner) [inline packages/core/src/template-compiler.ts]: This cache could be something that stats the file, and uses the cache if the file change. That might be a nice addition... thoughts @ef4

File: packages/core/src/template-compiler.ts [hunk 2]
```diff
@@ -215,7 +225,7 @@ export default class TemplateCompiler {
   @Memoize()
   private setup() {
     let syntax = loadGlimmerSyntax(this.params.compilerPath);
-    this.userPluginsCount += registerPlugins(syntax, this.params.plugins);
+    // this.userPluginsCount += registerPlugins(syntax, this.params.plugins);
     if (this.params.resolver) {
       let transform = this.params.resolver.astTransformer(this);
```
Comment (stefanpenner) [inline packages/core/src/template-compiler.ts]: TODO: this resolved transform needs to be incorporated in the new flow

**Comment by stefanpenner** [comment]
Some hbs transforms appear as classes, not pojos and using the new API cause:

```
  ember-paper/templates/components/paper-autocomplete-content.hbs: Class constructor EmberMaybeInElementAstTransform cannot be invoked without 'new'
            in /tmp/broccoli-6140i6g0b1tO5X2Y/out-03-funnel_addon_addon_template_files_ember_paper
            at embroider-template-compile-stage1
```

@rwjblue pointed out that ember should be wrapping these plugins via https://github.com/emberjs/ember.js/blob/bc435eb0a11868f7e0f8381f9a4bb2ba7728d9e8/packages/ember-template-compiler/lib/system/compile-options.ts#L55

Further investigation is required.

**Comment by stefanpenner** [comment]
After talking to rob, TL;DR we need to lift https://github.com/emberjs/ember.js/blob/bc435eb0a11868f7e0f8381f9a4bb2ba7728d9e8/packages/ember-template-compiler/lib/system/compile-options.ts#L55

**Comment by stefanpenner** [comment]
We could also use the same public API ember-cli uses to find and use ember-template-compiler.

A touch verbose:

```js
require(require('ember-source').absolutePaths.templateCompiler)
```

File: packages/core/src/template-compiler.ts [hunk 3]
```diff
@@ -36,18 +37,66 @@ interface GlimmerSyntax {
...
+      currentStat.mode === entry.stat.mode &&
+      currentStat.size === entry.stat.size &&
+      currentStat.mtime === entry.stat.mtime
+    ) {
+      console.count('HIT');
+      return entry.value;
+    }
+  }
+
+  console.count('MISS');
```
Comment (stefanpenner) [inline packages/core/src/template-compiler.ts]: TODO: remove

File: packages/core/src/template-compiler.ts [hunk 4]
```diff
@@ -36,18 +37,66 @@ interface GlimmerSyntax {
...
+  if (entry) {
+    let currentStat = statSync(templateCompilerPath);
+
+    // Let's ensure the template is still what we cached
+    if (
+      currentStat.mode === entry.stat.mode &&
+      currentStat.size === entry.stat.size &&
+      currentStat.mtime === entry.stat.mtime
+    ) {
+      console.count('HIT');
```
Comment (stefanpenner) [inline packages/core/src/template-compiler.ts]: TODO: remove

File: packages/core/src/template-compiler.ts [hunk 5]
```diff
@@ -215,12 +287,13 @@ export default class TemplateCompiler {
   @Memoize()
   private setup() {
     let syntax = loadGlimmerSyntax(this.params.compilerPath);
-    this.userPluginsCount += registerPlugins(syntax, this.params.plugins);
     if (this.params.resolver) {
       let transform = this.params.resolver.astTransformer(this);
       if (transform) {
+        // TODO: make this work
         syntax.registerPlugin('ast', transform);
```
Comment (stefanpenner) [inline packages/core/src/template-compiler.ts]: TODO: append this to userPlugins

File: packages/core/src/template-compiler.ts [hunk 6]
```diff
@@ -215,12 +287,13 @@ export default class TemplateCompiler {
...
         this.userPluginsCount++;
+        // throw new Error('EWUT');
```
Comment (stefanpenner) [inline packages/core/src/template-compiler.ts]: TODO: remove

File: packages/core/src/template-compiler.ts [hunk 7]
```diff
@@ -246,10 +319,13 @@ export default class TemplateCompiler {
       runtimeName = moduleName;
     }
 
+    let opts = this.syntax.defaultOptions({ contents, moduleName });
+    opts.plugins!.ast = [...this.params.plugins.ast!.slice().reverse(), ...opts.plugins!.ast!];
```
Comment (stefanpenner) [inline packages/core/src/template-compiler.ts]: TODO: make this not crappy

File: packages/core/src/template-compiler.ts [hunk 8]
```diff
@@ -287,8 +363,20 @@ export default class TemplateCompiler {
       // rather than slicing them off, we could choose instead to not call
       // syntax.defaultOptions, but then we lose some of the compatibility
       // normalization that it does on the user-provided plugins.
-      opts.plugins.ast = opts.plugins.ast.slice(0, this.userPluginsCount);
+      opts.plugins.ast = this.params.plugins.ast!
```
Comment (stefanpenner) [inline packages/core/src/template-compiler.ts]: TODO: make this not crappy

**Comment by ef4** [comment]
I think this was all superceded by https://github.com/embroider-build/embroider/pull/346

**Pull request closed without merging**